src/geometry/manifold/diffeomorph.lean

```diff
@@ -0,0 +1,81 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Nicolò Cavalleri.
+-/
+
+import geometry.manifold.times_cont_mdiff_map
+
+/-!
+# Diffeomorphisms
+This file implements diffeomorphisms.
+
+-/
+
+open_locale manifold
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{E' : Type*} [normed_group E'] [normed_space 𝕜 E']
+{F : Type*} [normed_group F] [normed_space 𝕜 F]
+{H : Type*} [topological_space H]
+{H' : Type*} [topological_space H']
+{G : Type*} [topological_space G]
+(I : model_with_corners 𝕜 E H) (I' : model_with_corners 𝕜 E' H')
+(J : model_with_corners 𝕜 F G)
+
+section diffeomorph
+
+variables (M : Type*) [topological_space M] [charted_space H M] [smooth_manifold_with_corners I M]
+(M' : Type*) [topological_space M'] [charted_space H' M'] [smooth_manifold_with_corners I' M']
+(N : Type*) [topological_space N] [charted_space G N] [smooth_manifold_with_corners J N]
+(n : with_top ℕ)
+
+/-- M and M' are diffeomorph with respect to I and I' -/
+@[protect_proj, nolint has_inhabited_instance]
+structure times_diffeomorph extends M ≃ M' :=
+(times_cont_mdiff_to_fun  : times_cont_mdiff I I' n to_fun)
+(times_cont_mdiff_inv_fun : times_cont_mdiff I' I n inv_fun)
+
+/-- A `diffeomorph` is just a smooth `times_diffeomorph`. -/
+@[reducible] def diffeomorph := times_diffeomorph I I' M M' ⊤
+
+infix ` ≃ₘ `:50 := times_diffeomorph _ _
+localized "notation M ` ≃ₘ^ `n `⟮` I `,` J `⟯` N := times_diffeomorph I J M N n" in manifold
```

Comment:
Suggestions for better notation are welcome (I still would like a notation here but I'd be happy to improve the one I chose)